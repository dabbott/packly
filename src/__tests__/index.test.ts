import { Node, path, Volume } from 'imfs'
import { parse, serialize } from 'parse5'
import { bundle, BundleOptions, RequestOptions } from '..'

const createOptions = (volume: Node<string>): BundleOptions => ({
  entry: '/index.html',
  request: ({ url, origin }: RequestOptions) => {
    if (/^(https?)?\/\//.test(url)) return undefined

    return Volume.readFile(
      volume,
      origin && !url.startsWith('/') ? path.join(origin, '..', url) : url
    )
  },
})

it('identity', () => {
  const volume = Volume.create<string>()

  const html = `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>My page</title>
  </head>
  <body>
    <p>Hello, world!</p>
  </body>
</html>`

  Volume.writeFile(volume, `/index.html`, html)

  const result = bundle(createOptions(volume))

  expect(result).toEqual(serialize(parse(html)))
})

describe('styles', () => {
  it('inlines a style', () => {
    const volume = Volume.create<string>()

    const html = `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>My page</title>
      <link rel="stylesheet" href="style.css">
    </head>
    <body>
      <p>Hello, world!</p>
    </body>
  </html>`

    const css = `body { background: #222; }`

    Volume.writeFile(volume, `/index.html`, html)
    Volume.writeFile(volume, `/style.css`, css)

    const result = bundle(createOptions(volume))

    expect(result).toMatchSnapshot()
  })

  it('ignores a style', () => {
    const html = `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>My page</title>
      <link rel="stylesheet" href="style.css">
    </head>
    <body>
      <p>Hello, world!</p>
    </body>
  </html>`

    const result = bundle({
      entry: '/index.html',
      request: ({ url }: RequestOptions) => {
        if (url === '/index.html') return html
        return undefined
      },
    })

    expect(result).toEqual(serialize(parse(html)))
  })
})

describe('scripts', () => {
  it('inlines a script', () => {
    const volume = Volume.create<string>()

    const html = `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>My page</title>
      <script src="main.js"></script>
    </head>
    <body>
      <p>Hello, world!</p>
    </body>
  </html>`

    const js = `console.log('Hello, world!');`

    Volume.writeFile(volume, `/index.html`, html)
    Volume.writeFile(volume, `/main.js`, js)

    const result = bundle(createOptions(volume))

    expect(result).toMatchSnapshot()
  })

  it('ignores a script', () => {
    const html = `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>My page</title>
      <script src="main.js"></script>
    </head>
    <body>
      <p>Hello, world!</p>
    </body>
  </html>`

    const result = bundle({
      entry: '/index.html',
      request: ({ url }: RequestOptions) => {
        if (url === '/index.html') return html
        return undefined
      },
    })

    expect(result).toEqual(serialize(parse(html)))
  })
})
