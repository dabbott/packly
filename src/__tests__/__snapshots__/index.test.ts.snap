// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scripts inlines a script 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
      <meta charset=\\"utf-8\\">
      <title>My page</title>
      <script>console.log('Hello, world!');</script>
    </head>
    <body>
      <p>Hello, world!</p>
    
  </body></html>"
`;

exports[`styles inlines a style 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
      <meta charset=\\"utf-8\\">
      <title>My page</title>
      <style>body { background: #222; }</style>
    </head>
    <body>
      <p>Hello, world!</p>
    
  </body></html>"
`;
